graphql_compiler/tests/test_compiler.py

```diff
@@ -2724,3 +2724,60 @@ def test_coercion_filters_and_multiple_outputs_within_fold_traversal(self):
         '''
 
         check_test_data(self, test_data, expected_match, expected_gremlin)
+
+    def test_optional_and_fold(self):
+        test_data = test_input_data.optional_and_fold()
+
+        expected_match = '''
+            SELECT
+                Animal___1.name AS `animal_name`,
+                $Animal___1___out_Animal_ParentOf.name AS `child_names_list`,
+                if(eval("(Animal__in_Animal_ParentOf___1 IS NOT null)"),
+                    Animal__in_Animal_ParentOf___1.name,
+                    null
+                ) AS `parent_name`
+            FROM (
+                MATCH {{
+                    class: Animal,
+                    as: Animal___1
+                }}.in('Animal_ParentOf') {{
+                    optional: true,
+                    as: Animal__in_Animal_ParentOf___1
+                }}
+                RETURN $matches
+            )
+            LET
+                $Animal___1___out_Animal_ParentOf = Animal___1.out("Animal_ParentOf").asList()
+        '''
+        expected_gremlin = '''
+            g.V('@class', 'Animal')
+            .as('Animal___1')
+                .ifThenElse{it.in_Animal_ParentOf == null}{null}{it.in('Animal_ParentOf')}
+                .as('Animal__in_Animal_ParentOf___1')
+            .optional('Animal___1')
+            .as('Animal___2')
+            .transform{it, m -> new com.orientechnologies.orient.core.record.impl.ODocument([
+                animal_name: m.Animal___1.name,
+                child_names_list: (
+                    (m.Animal___2.out_Animal_ParentOf == null) ?
+                        [] :
+                        (m.Animal___2.out_Animal_ParentOf.collect{entry -> entry.inV.next().name})
+                ),
+                parent_name: (
+                    (m.Animal__in_Animal_ParentOf___1 != null) ?
+                        m.Animal__in_Animal_ParentOf___1.name : null
+                )
+            ])}
+        '''
+
+        check_test_data(self, test_data, expected_match, expected_gremlin)
+
+    def test_fold_and_optional(self):
+        test_data = test_input_data.fold_and_optional()
+
+        expected_match = '''
```

Comment:
did you mean to leave these empty?

Reply:
Yes, this query actually causes a compilation error.

Reply:
I'm not sure I understand -- compilation error as in a `raise` kind of error, or just a buggy compilation?

Do both of these testcases currently fail, or just one of them?

Reply:
The first one produces incorrect MATCH because of revisits.
The second one raises an `AssertionError`

Reply:
Can you post all this info as a comment in the PR, including the specifics on why the MATCH is wrong and what the AssertionError is? This is a bug report, and it should be as detailed as possible so that other people that run into it can correctly triage and assess it -- and this will also help with fixing it.